ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Unary.hs

```diff
@@ -96,316 +65,442 @@ import           Ouroboros.Consensus.HardFork.Combinator.Mempool
-                    -> ConsensusConfig (BlockProtocol (HardForkBlock '[b]))
-                    -> ConsensusConfig (BlockProtocol b)
-projConsensusConfig ei =
-      completeConsensusConfig (Proxy @(BlockProtocol b)) ei
-    . unwrapPartialConsensusConfig
-    . hd
-    . getPerEraConsensusConfig
-    . hardForkConsensusConfigPerEra
-
-injConsensusConfig :: forall b. NoHardForks b
-                   => History.EraParams
-                   -> ConsensusConfig (BlockProtocol b)
-                   -> ConsensusConfig (BlockProtocol (HardForkBlock '[b]))
-injConsensusConfig eraParams cfg = HardForkConsensusConfig {
-      hardForkConsensusConfigK      = protocolSecurityParam cfg
-    , hardForkConsensusConfigShape  = History.singletonShape eraParams
-    , hardForkConsensusConfigPerEra = PerEraConsensusConfig $
-           WrapPartialConsensusConfig (toPartialConsensusConfig (Proxy @b) cfg)
-        :* Nil
-    }
-
-projLedgerState :: LedgerState (HardForkBlock '[b]) -> LedgerState b
-projLedgerState =
-      State.currentState
-    . Telescope.fromTZ
-    . getHardForkState
-    . getHardForkLedgerState
+instance Isomorphic WrapTipInfo where
+  project = defaultProjectNS
+  inject  = defaultInjectNS
 
-injLedgerState :: LedgerState b -> LedgerState (HardForkBlock '[b])
-injLedgerState =
-      HardForkLedgerState
-    . HardForkState
-    . Telescope.TZ
-    . State.Current History.initBound
+{-------------------------------------------------------------------------------
+  Hash
+-------------------------------------------------------------------------------}
+
+instance Isomorphic WrapHeaderHash where
+  project = aux
+    where
+      aux :: forall blk. ConvertRawHash blk
+          => WrapHeaderHash (HardForkBlock '[blk]) -> WrapHeaderHash blk
+      aux = WrapHeaderHash
+          . fromRawHash (Proxy @blk) . getOneEraHash
+          . unwrapHeaderHash
+
+  inject = aux
+    where
+      aux :: forall blk. ConvertRawHash blk
+          => WrapHeaderHash blk -> WrapHeaderHash (HardForkBlock '[blk])
+      aux = WrapHeaderHash
+          . OneEraHash . toRawHash (Proxy @blk)
+          . unwrapHeaderHash
+
+instance Isomorphic ChainHash where
+  project = aux
+    where
+      aux :: forall blk. NoHardForks blk
+          => ChainHash (HardForkBlock '[blk]) -> ChainHash blk
+      aux GenesisHash   = GenesisHash
+      aux (BlockHash h) = BlockHash (project' (Proxy @(WrapHeaderHash blk)) h)
+
+  inject = aux
+    where
+      aux :: forall blk. NoHardForks blk
+          => ChainHash blk -> ChainHash (HardForkBlock '[blk])
+      aux GenesisHash   = GenesisHash
+      aux (BlockHash h) = BlockHash (inject' (Proxy @(WrapHeaderHash blk)) h)
+
+{-------------------------------------------------------------------------------
+  Config
+
+  This is a bit trickier due to the partial configs.
+-------------------------------------------------------------------------------}
+
+instance Isomorphic TopLevelConfig where
```

Comment:
Separate instances for ledger config / consensus config?

Reply:
Can't be done. Added a comment:

```haskell
-- | Projection/injection for 'TopLevelConfig'
--
-- NOTE: We do not define one for 'LedgerConfig' or 'ConsensusConfig', since
-- we need the 'EraParams' for their injections, which we can only derive if
-- we have the top-level config.
```